### PR TITLE
testing improvements

### PR DESCRIPTION
**Pull request by turadg**
## Description

https://github.com/Agoric/agoric-sdk/pull/5766 ran into a problem with `test.todo` not being available in ses-ava. 524e7de69408414abcfed42d6468b01953f9897b adds it. (fyi @erights )

I also got fed up with this noise in test output. 295c8aadf1840e1a315ab0354c83659e161db369 removes it.
Screen Shot 2022-07-12 at 7 55 13 AM

### Security Considerations

none

### Documentation Considerations

--

### Testing Considerations

Better

### CODE REVIEW

File: patches/@endo+ses-ava+0.2.27.patch
Comment (kriskowal) [inline patches/@endo+ses-ava+0.2.27.patch]:
```suggestion
+ * @returns {TestInterface}
```

File: patches/@endo+ses-ava+0.2.27.patch
Comment (kriskowal) [inline patches/@endo+ses-ava+0.2.27.patch]: Are we confident that our emulation satisfies the entire ava interface now? Or, is this a convenient fiction that improves the debug experience?

**Review by kriskowal — APPROVED** [review]

File: patches/@endo+ses-ava+0.2.27.patch
Comment (turadg) [reply patches/@endo+ses-ava+0.2.27.patch]: You're right we don't. That's why this still returns `TestInterface`. But the `avaTest` param is indeed from Ava.

Marking this as Draft because my refactor broke some case.

**Comment by turadg** [comment]
@kriskowal why do we have a whitelist of what to allow through? shouldn't it be an "override" list that's passed over after copying everything? IIUC "ses-ava" package is to provide SES-happy console output, not for actual security of the test run.

**Comment by kriskowal** [comment]
> @kriskowal why do we have a whitelist of what to allow through? shouldn't it be an "override" list that's passed over after copying everything? IIUC "ses-ava" package is to provide SES-happy console output, not for actual security of the test run.

I should in this case defer to @erights, but this may have been born of abundance-of-caution or force-of-security-habit and eligible for a rethink.

**Comment by turadg** [comment]
I looked into the [open issues for ses-ava](https://github.com/endojs/endo/issues?q=is%3Aissue+is%3Aopen+ses-ava) and these would be solved by having the "override" list instead of an "allow" list:
https://github.com/endojs/endo/issues/645
https://github.com/endojs/endo/issues/647

But I don't understand why this fence was put here! @erights ?

I also don't know the symptoms of not using `ses-ava` at all. Do they persist in Ava 4?

**Comment by turadg** [comment]
Also, since the `prepare-test-env` is mutating the env, why not mutate Ava `test` itself? Status quo is we have these `prepare-test-env-ava.js` files that have both side-effects AND exports. Because they have side-effects they have to go before other imports. But then when they use relative paths our lint rules complain that it should be lower, so we have to suppress:
```
/* eslint-disable import/order */
import { test } from '../tools/prepare-test-env-ava.js';
```
There are 36 of those.

How about this instead?
```
import '../tools/prepare-test-env-ava.js';

import test from 'ava';
```

File: packages/run-protocol/test/test-gov-collateral.js
Comment (erights) [inline packages/run-protocol/test/test-gov-collateral.js:589]: What is the whole `test` vs `anyTest` dance at the beginning of this file about? I don't get it. Can we simplify this to follow the pattern in the rest of this PR?

File: packages/SwingSet/test/test-message-patterns.js
Comment (erights) [inline packages/SwingSet/test/test-message-patterns.js:6]: I'm fine with this change, but confused about your intentions for this PR. At https://github.com/Agoric/agoric-sdk/pull/5774#issuecomment-1186236576 you seem to suggest moving in exactly the opposite direction. Preferring the status quo over the change you're making here.

Regarding the need for the `eslint-disable`, could we instead have all of these be `@foo/something` imports, probably from some (new?) package in endo?

File: packages/solo/test/test-home.js
Comment (erights) [inline packages/solo/test/test-home.js:3]: Again, seems to go in the opposite direction to your suggestion at https://github.com/Agoric/agoric-sdk/pull/5774#issuecomment-1186236576 . I actually prefer the new form above over your suggestion there, but I don't understand which direction you want this PR to take things.

I mildy prefer the form above simply because it is shorter and bundles the two relevant things together. As separate imports, the dependency between them is less clear, and more of a refactoring hazard.

Why are you using the `@foo/something` for here, thereby avoiding the need for `eslint-disable`, but not in other places?

File: packages/ui-components/test/display/test-display.js
Comment (erights) [inline packages/ui-components/test/display/test-display.js:1]: Why is this `eslint-disable` needed?

File: packages/zoe/src/contractFacet/zcfZygote.js
Comment (erights) [inline packages/zoe/src/contractFacet/zcfZygote.js:322]: Happy to see this go. But do we know why it was there?

File: packages/zoe/test/unitTests/contracts/loan/test-addCollateral.js
Comment (erights) [inline packages/zoe/test/unitTests/contracts/loan/test-addCollateral.js:1]: Surprised to see this in a PR from you ;)

Likewise with several other occurrences below.

File: patches/@endo+ses-ava+0.2.27.patch
Comment (erights) [inline patches/@endo+ses-ava+0.2.27.patch:4]: Oh geez, an endo patch. Is there (or will there be) a corresponding endo PR that, once integrated, would enable us to remove this patch file? I would rather first review this as an endo PR before reviewing it as an endo patch to agoric-sdk.

File: patches/@endo+ses-ava+0.2.27.patch
Comment (erights) [inline patches/@endo+ses-ava+0.2.27.patch]: It looks like the following terminology problem was already there. But this PR might still be a good opportunity to fix it.

The term `thunk` here seems to refer to the `func` parameter below. If true, they should both use the same name. The CS term `thunk` that I'm familiar with refers only to a function/closure with zero arguments, which `func` is not. So if they are referring to the same thing, `func` is a better choice.

Hmm. I think I see how this terminology problem happened (and I think I may have perpetrated it). The `func` together with the `args` may have previously been packaged together as one `thunk` argument that was invoked with no args. But that always created a yet deeper stack exactly in the place where we're trying to generate good diagnostic info. So by separating them in this way, we're able to make the diagnostic stacks clearer. Yeah, I think I remember doing this. Sorry I did not clean up the comment.

File: patches/@endo+ses-ava+0.2.27.patch
Comment (erights) [inline patches/@endo+ses-ava+0.2.27.patch]: I don't understand replacing the specific type `TesterFunc` with a generic type parameter. The returned `augmented` func assumes that the `testerFunc`'s second parameter is an `ImplFunc`. The `augmented` function itself is supposed to be a compat replacement for `testerFunc` only under this assumption, since it has its own `implFunc` parameter, which it replaces with `testFuncWrapper`.

If `testerFunc` is anything else, then `wrapTester(testerFunc, ...)` would be wrong. The method names of the `overrideList`/`testerMethodsWhitelist` are supposed to enumerate only methods of `test` which conform to these assumptions about `TesterFunc`.

File: patches/@endo+ses-ava+0.2.27.patch
Comment (erights) [inline patches/@endo+ses-ava+0.2.27.patch:144]: Is this needed because you're (inappropriately?) using a generic type parameter rather than `TesterFunc`?

File: patches/@endo+ses-ava+0.2.27.patch
Comment (erights) [inline patches/@endo+ses-ava+0.2.27.patch:143]: An interesting choice. This copies all the old properties of `testerFunc` onto `augmented` whether or not they have been augmented. Whereas the original sought to omit methods that should have been augmented but were not, this will preserve the original even though we had failed to augment it. This is plausible, but we should discuss.

File: patches/@endo+ses-ava+0.2.27.patch
Comment (erights) [inline patches/@endo+ses-ava+0.2.27.patch:148]: Do you plan to check before this PR is R4R?

File: patches/@endo+ses-ava+0.2.27.patch
Comment (erights) [inline patches/@endo+ses-ava+0.2.27.patch]: This `@template` line has more syntax than I recognize. What does it mean?

File: patches/@endo+ses-ava+0.2.27.patch
Comment (erights) [inline patches/@endo+ses-ava+0.2.27.patch:187]: IIUC, you're still returning a new augmented `sesAvaTest` rather than mutating the original `avaTest`. Yes? I don't mind. But didn't your PR comment suggest mutating the original in place?

File: patches/@endo+ses-ava+0.2.27.patch
Comment (erights) [inline patches/@endo+ses-ava+0.2.27.patch:191]: If it is exported, it should be hardened. Unless we're purposely avoiding this because this code runs before `lockdown`? Are we? Does it?

File: patches/@endo+ses-ava+0.2.27.patch
Comment (erights) [inline patches/@endo+ses-ava+0.2.27.patch:252]: Yeah, this (and its dependents) are the ones I am skeptical you can omit while still using types to describe the augmentation.

**Review by erights — CHANGES_REQUESTED** [review]
Since I see @kriskowal already approved, I'm changing my comments to Request changes as an interlock, even though it is still not R4R

**Comment by erights** [comment]
> @kriskowal why do we have a whitelist of what to allow through? shouldn't it be an "override" list that's passed over after copying everything? IIUC "ses-ava" package is to provide SES-happy console output, not for actual security of the test run.

Yeah, it isn't for security. It is to keep track of which methods we've examined to determine whether to augment them or not. Perhaps we should have two whitelists: The ones we know we want to augment, and the ones we know we don't. It is an interesting policy choice what to do with the remainder: Do we preserve their existing behavior (as your code does) until we decide, or do we omit them until we decide?

**Comment by erights** [comment]
> Also, since the `prepare-test-env` is mutating the env, why not mutate Ava `test` itself? Status quo is we have these `prepare-test-env-ava.js` files that have both side-effects AND exports. Because they have side-effects they have to go before other imports. But then when they use relative paths our lint rules complain that it should be lower, so we have to suppress:
> 
> ```
> /* eslint-disable import/order */
> import { test } from '../tools/prepare-test-env-ava.js';
> ```
> 
> There are 36 of those.
> 
> How about this instead?
> 
> ```
> import '../tools/prepare-test-env-ava.js';
> 
> import test from 'ava';
> ```

Whether or not we mutate `test` in place, shouldn't we still prefer the former to the latter because it is shorter (less big of a deal) and bundles related issues together so it is less of a refactoring hazard?

Also, wouldn't the first line of the latter still need to come first? If it does and still uses the '`../tools/...'` form you show above, wouldn't it still need the `eslint-disable`?

**Comment by erights** [comment]
> I also don't know the symptoms of not using ses-ava at all. Do they persist in Ava 4?

The consequence is that the seeming console output is neither from the real console nor from the replacement that `lockdown` installs instead. Thus it lacks the crucial diagnostic info that's especially important when we're iterating the development of our test.

**Comment by turadg** [comment]
> Yeah, it isn't for security.

Ok good. So the purpose of the "ses-ava" package isn't for security, just to accommodate SES's console handling when testing with Ava.

> It is to keep track of which methods we've examined to determine whether to augment them or not. Perhaps we should have two whitelists: The ones we know we want to augment, and the ones we know we don't. It is an interesting policy choice what to do with the remainder: Do we preserve their existing behavior (as your code does) until we decide, or do we omit them until we decide?

I can't think of any reason to omit any part of the Ava test framework. The user shouldn't have to lose functionality because of SES's console requirements. (The lack of `.todo` was a surprise which broke my build and my focus.)  For runtime, one list suffices: modify these methods so work better in SES. The rest should pass through. If there is any value in keeping track of "we looked into this one and it doesn't need augmentation", that's a documentation concern. A list like that in a comment above the override list would be fine by me, though I don't see the value in that. I expect whenever there's a problem that it'll get reported and then belong on the override list. If something is known to need overriding and that's not working yet, that could go into the override list, commented out with an explanation (and ideally a link to a ticket).

**Comment by turadg** [comment]
> > ```
> > /* eslint-disable import/order */
> > import { test } from '../tools/prepare-test-env-ava.js';
> > ```
> >   
> > How about this instead?
> > ```
> > import '../tools/prepare-test-env-ava.js';
> > 
> > import test from 'ava';
> > ```
> 
> Whether or not we mutate `test` in place, shouldn't we still prefer the former to the latter because it is shorter (less big of a deal) and bundles related issues together so it is less of a refactoring hazard?

I think those are both offset by the value of consistent Ava tests, regardless of whether SES env is prepared. What's "related" is "prepare env" and when you prepare env or not, it should not change the ergonomics of writing a test. By bundling shims which have side effects with exports of symbols to use, I argue that this is more of a refactoring hazard.

> Also, wouldn't the first line of the latter still need to come first? If it does and still uses the '`../tools/...'` form you show above, wouldn't it still need the `eslint-disable`?

Good question. The way the rule works, if you import without a symbol it's assumed to have side effects so it's exempt from ordering rules.

Note that this PR isn't requesting the change. I asked to understand why it's like this in the first place and to see if we can agree on the proposed way. (I've since seen in PR history of Endo that it used to be the way I'm proposing.) If we do, I'll file a ticket to do that after MN-1.

But my immediate goal with this PR is to make SES-Ava conform to the Ava API.

File: patches/@endo+ses-ava+0.2.27.patch
Comment (turadg) [reply patches/@endo+ses-ava+0.2.27.patch:191]: I don't know why it was commented out but that seems plausible. I don't have time to research further but I can restore the line if will get this approved :)

File: patches/@endo+ses-ava+0.2.27.patch
Comment (turadg) [reply patches/@endo+ses-ava+0.2.27.patch:187]: Correct, in this PR not mutating the original. [Clarified](https://github.com/Agoric/agoric-sdk/pull/5774#issuecomment-1186539828)

File: patches/@endo+ses-ava+0.2.27.patch
Comment (turadg) [reply patches/@endo+ses-ava+0.2.27.patch]: `@template T` you're familiar with. The `{Function}` means T has to be a subtype of `Function`. That's so TS knows that the calls below are valid. The string after `T` is a description for display.

File: patches/@endo+ses-ava+0.2.27.patch
Comment (turadg) [reply patches/@endo+ses-ava+0.2.27.patch:148]: Nope. Ava 4 is on hold to at least after MN-1 https://github.com/Agoric/agoric-sdk/issues/4895

File: packages/zoe/src/contractFacet/zcfZygote.js
Comment (turadg) [reply packages/zoe/src/contractFacet/zcfZygote.js:322]: Nope. Given how old it is I would guess it was added around when test jigs were new, to notify the dev about it.

File: packages/zoe/test/unitTests/contracts/loan/test-addCollateral.js
Comment (turadg) [reply packages/zoe/test/unitTests/contracts/loan/test-addCollateral.js:1]: ha, well, a few reasons:
- out of scope for PR
- not a priority before MN-1 to add types to code we're not working in
- types in tests are only for the test writer since they don't provide any additional confidence in production

File: patches/@endo+ses-ava+0.2.27.patch
Comment (turadg) [reply patches/@endo+ses-ava+0.2.27.patch:4]: > Is there (or will there be) a corresponding endo PR that, once integrated, would enable us to remove this patch file?

Yes there will be. I think that goes for every patch file Agoric can push upstream. So I think the existence of this file is documentation enough.

> I would rather first review this as an endo PR before reviewing it as an endo patch to agoric-sdk.

My job right now is to ship MN-1. That ses-ava breaks the Ava API is slowing me down. Effort fixing it in Endo proper should not be spent until after shipping.

File: packages/ui-components/test/display/test-display.js
Comment (turadg) [reply packages/ui-components/test/display/test-display.js:1]: Because the import in the next line isn't declared in the package dependencies.

File: packages/solo/test/test-home.js
Comment (turadg) [reply packages/solo/test/test-home.js:3]: I made this change to be consistent with the other code.  I do think we'd be better off with `import test from 'ava'` but that's [out of scope](https://github.com/Agoric/agoric-sdk/pull/5774#issuecomment-1186539828)).

> Why are you using the @foo/something for here … but not in other places?

I didn't change any import paths so not really my decision. But I think the relative ones are because what's imported is coming from the same package.

File: packages/SwingSet/test/test-message-patterns.js
Comment (turadg) [reply packages/SwingSet/test/test-message-patterns.js:6]: > Regarding the need for the eslint-disable, could we instead have all of these be @foo/something imports, probably from some (new?) package in endo?

Not sure I follow but I'm sure it's out of scope for this PR. I'm just trying to make the `test` function work consistently. While I would like to refactor it out of the `prepare-` files that's for another day.

File: packages/run-protocol/test/test-gov-collateral.js
Comment (turadg) [reply packages/run-protocol/test/test-gov-collateral.js:589]: It's a cast so that `t.context` is statically typed. E.g. in this file `t.context.zoe` is known to be `ZoeService`.

Screen Shot 2022-07-17 at 8 10 33 AM
.

File: patches/@endo+ses-ava+0.2.27.patch
Comment (turadg) [reply patches/@endo+ses-ava+0.2.27.patch]: Confident now that it satisfies because it passes through whatever it doesn't override. I typed the param as generic so that it wraps whatever version of Ava was passed in.

File: patches/@endo+ses-ava+0.2.27.patch
Comment (turadg) [reply patches/@endo+ses-ava+0.2.27.patch]: If you want to now feel free to submit a suggestion through Github UI or push a commit on the branch.

File: patches/@endo+ses-ava+0.2.27.patch
Comment (turadg) [reply patches/@endo+ses-ava+0.2.27.patch]: I went with any `Function` because it works. I don't see the value in the additional constraints. If this is a blocker let me know.

File: patches/@endo+ses-ava+0.2.27.patch
Comment (turadg) [reply patches/@endo+ses-ava+0.2.27.patch:144]: Possibly. The error the cast is covering up,
```
Type '(title: any, implFunc: any, ...otherArgs: any[]) => any' is not assignable to type 'T'.
  '(title: any, implFunc: any, ...otherArgs: any[]) => any' is assignable to the constraint of type 'T', but 'T' could be instantiated with a different subtype of constraint 'Function'.
```
But again I don't see the value in being precise in this particular code. Let me know if it's a requirement to merge into agoric-sdk.

File: patches/@endo+ses-ava+0.2.27.patch
Comment (turadg) [reply patches/@endo+ses-ava+0.2.27.patch:143]: I assert this is the ideal behavior. The job of the wrapper is to fix console output where we know it's a problem. It shouldn't break anything it doesn't know about.

File: patches/@endo+ses-ava+0.2.27.patch
Comment (turadg) [reply patches/@endo+ses-ava+0.2.27.patch:252]: The important types here are what `wrapTest` returns and it's as simple as "whatever it got" because it shouldn't be changing the API of Ava.

File: packages/run-protocol/test/test-gov-collateral.js
Comment (erights) [reply packages/run-protocol/test/test-gov-collateral.js:589]: Wow, that's good to know. Resolving.

File: packages/SwingSet/test/test-message-patterns.js
Comment (erights) [reply packages/SwingSet/test/test-message-patterns.js:6]: Gotcha. Resolving.

File: packages/solo/test/test-home.js
Comment (erights) [reply packages/solo/test/test-home.js:3]: ok. Resolving.

File: packages/ui-components/test/display/test-display.js
Comment (erights) [reply packages/ui-components/test/display/test-display.js:1]: Should it be?

File: packages/zoe/src/contractFacet/zcfZygote.js
Comment (erights) [reply packages/zoe/src/contractFacet/zcfZygote.js:322]: ok. Resolving.

File: packages/zoe/test/unitTests/contracts/loan/test-addCollateral.js
Comment (erights) [reply packages/zoe/test/unitTests/contracts/loan/test-addCollateral.js:1]: It is interesting to see how our coding standards for test code increasingly differs from our coding standards for production code. Another is that we should not impose our top-level-await rule on test code. Will we maintain divergent lint settings?

In any case, for this PR no change suggested. Resolving.

File: patches/@endo+ses-ava+0.2.27.patch
Comment (erights) [reply patches/@endo+ses-ava+0.2.27.patch:4]: Sounds good to me. How do we remember to get back to this after MN-1? Is there an issue?

Resolving.

File: patches/@endo+ses-ava+0.2.27.patch
Comment (erights) [reply patches/@endo+ses-ava+0.2.27.patch]: It is not a blocker. I recognize that you understand our types much better than I. So when I'm puzzled by your use of types, I'd at least like to learn from you. 

Here, I am genuinely confused about why you think this change is an improvement. The code being described by these types relies on the assumption that the second parameter of the testerFunc is an ImplFunc, in order to replace this ImplFunc by wrapping. If we didn't have a type language to describe it, I'd feel compelled to write prose describing that. But we do have a type language that is good at saying things like this. Why not use it?

I suppose the blocker is that if we do not explain this constraint using the type language, then we do need explicit prose describing it well. Why wouldn't we?

File: patches/@endo+ses-ava+0.2.27.patch
Comment (erights) [reply patches/@endo+ses-ava+0.2.27.patch:144]: I don't care about the type error. I do care about clarity to programmers reading and maintaining this code.

File: patches/@endo+ses-ava+0.2.27.patch
Comment (erights) [reply patches/@endo+ses-ava+0.2.27.patch:143]: I am ok with this rationale. Resolving.

File: patches/@endo+ses-ava+0.2.27.patch
Comment (erights) [reply patches/@endo+ses-ava+0.2.27.patch:148]: thanks. Resolving.

File: patches/@endo+ses-ava+0.2.27.patch
Comment (erights) [reply patches/@endo+ses-ava+0.2.27.patch]: Ah. Although I knew that extra tokens after jsdoc/ts declarations are assumed to be prose descriptions, somehow that did not occur to me here. It seemed like more type notation that I didn't know. Thanks for clarifying. Resolving.

File: patches/@endo+ses-ava+0.2.27.patch
Comment (erights) [reply patches/@endo+ses-ava+0.2.27.patch:191]: Please restore and uncomment out the harden call.

File: patches/@endo+ses-ava+0.2.27.patch
Comment (erights) [reply patches/@endo+ses-ava+0.2.27.patch:252]: I do not understand your reply.

**Comment by erights** [comment]
> If there is any value in keeping track of "we looked into this one and it doesn't need augmentation", that's a documentation concern. 

I am ok with this, and with your other points above that I did not comment on.

File: patches/@endo+ses-ava+0.2.27.patch
Comment (erights) [reply patches/@endo+ses-ava+0.2.27.patch]: Manually editing a patch seems a bit much, so https://github.com/endojs/endo/pull/1234 instead ;)

Resolving here.

File: packages/ui-components/test/display/test-display.js
Comment (turadg) [reply packages/ui-components/test/display/test-display.js:1]: Out of scope for this PR but I don't think necessarily so. That would be most expedient.

It's also worth considering that whoever added the dependence on `@agoric/swingset-vat/tools/prepare-test-env-ava.js` didn't want `ui-components` to depend on `swingset-vat` even for dev. The alternative is to put that file somewhere else. It's dependency tree can be satisfied by Endo and `vat-data`.

File: packages/zoe/test/unitTests/contracts/loan/test-addCollateral.js
Comment (turadg) [reply packages/zoe/test/unitTests/contracts/loan/test-addCollateral.js:1]: > Will we maintain divergent lint settings?

It's managed by a rule tightening for production code:
https://github.com/Agoric/agoric-sdk/blob/9410a8500f6a30906c9e3359ece0e5d8dfd965ad/.eslintrc.cjs#L73-L91

> our coding standards for test code increasingly differs from our coding standards for production code

Put more exactly, some requirements for production are relaxed for non-production. I think this is appropriate.

File: patches/@endo+ses-ava+0.2.27.patch
Comment (turadg) [reply patches/@endo+ses-ava+0.2.27.patch:4]: I would think the presence of the patchfile would be a sufficient reminder but I've https://github.com/endojs/endo/issues/1235

File: patches/@endo+ses-ava+0.2.27.patch
Comment (turadg) [reply patches/@endo+ses-ava+0.2.27.patch]: > confused about why you think this change is an improvement

I don't necessarily think every line is an improvement. I think the overall PR is an improvement.

Is it a regression? It's a loss of type information but it's now _less wrong_ because not every tester function takes an impl. E.g. `test.todo` https://github.com/endojs/endo/issues/645

> we do have a type language that is good at saying things like this. Why not use it?

Because of the cost/benefit of making it accurate and not misleading. With this PR the costs of not doing it have been going up so I've gone ahead and re-added TesterFunc and made `implFunc` optional.

File: patches/@endo+ses-ava+0.2.27.patch
Comment (turadg) [reply patches/@endo+ses-ava+0.2.27.patch:144]: > care about clarity to programmers reading and maintaining this code.

Here here! :)

File: patches/@endo+ses-ava+0.2.27.patch
Comment (turadg) [reply patches/@endo+ses-ava+0.2.27.patch:252]: Basically https://github.com/Agoric/agoric-sdk/pull/5774/#discussion_r923405025

File: packages/ui-components/test/display/test-display.js
Comment (erights) [reply packages/ui-components/test/display/test-display.js:1]: Good. Resolving.

File: patches/@endo+ses-ava+0.2.27.patch
Comment (erights) [reply patches/@endo+ses-ava+0.2.27.patch:191]: Thx. Resolving

File: patches/@endo+ses-ava+0.2.27.patch
Comment (erights) [reply patches/@endo+ses-ava+0.2.27.patch]: Aha! I think I understand our misunderstanding. You say 
> not every tester function takes an impl. E.g. `test.todo`

But I omitted `todo` from the whitelist, and I see you still omit it from the `overrideList`. Then it occurred to me: I am using `TesterFunc` only to describe the methods on this list. You (and possibly TS itself?) are interpreting it as a description of all methods on `test`, which it clearly is not. If `todo` isn't described by `TesterFunc` then it should not be on this list anyway.

Making `implFunc` optional seems like a solution only because of our misunderstanding.

How should this be expressed in the type system? I thought I had done that successfully by enumerating in `TesterProperties` exactly the methods I put into the list. `todo` is absent both from the list and from the `TesterProperties` type. 

(Having the `TesterProperties` type track the list is a redundancy that creates a refactoring hazards, which I would also like to avoid if possible. But let's take that question separately.)

File: patches/@endo+ses-ava+0.2.27.patch
Comment (erights) [reply patches/@endo+ses-ava+0.2.27.patch:252]: Good. Answered there.

Btw, a github weirdness that seems new to me: When I click or even hover over the link above I see that github rendered the markdown without the url fragment, thereby linking only to the PR as a whole, which confused me. So I selected "Edit" from the triple-dots of this comment to see the source, then copied the url source for this link into the url bar of a separate browser tab, and viola, I saw the comment you refer to.

Comment (turadg) [reply patches/@endo+ses-ava+0.2.27.patch:252]: ah, that is weird. seems like a bug

File: patches/@endo+ses-ava+0.2.27.patch
Comment (turadg) [reply patches/@endo+ses-ava+0.2.27.patch]: Yeah, that's it. Do statically type this we'd have to annotate that `todo` is different from the others. I don't think this is worth it. I'm just trying to make `test` work so I can use it to develop. This PR makes it work. I greatly value code maintainability and I think this is maintainable enough. The type system doesn't have to cover every expression.

I don't totally know the refactoring hazards you're trying to avoid but I suspect that whatever time we're saving on that end we're simply spending now, when time is more expensive.

Comment (erights) [reply patches/@endo+ses-ava+0.2.27.patch]: I hear the frustration. If this were a PR against genuine source code I would feel uncomfortable rushing approval on this basis. However, here I'll take advantage of the fact that the changes I object to are on a patch. When it is time to make a corresponding change against the original sources on endo, please make me a reviewer and let's resolve this misunderstanding then.

**Review by erights — APPROVED** [review]
Approving in a rush as explained at https://github.com/Agoric/agoric-sdk/pull/5774#discussion_r923832098

File: patches/@endo+ses-ava+0.2.27.patch
Comment (turadg) [reply patches/@endo+ses-ava+0.2.27.patch]: Sounds like a good plan. =)

**Pull request merged**
